### DISCUSSION TITLE
Allow more comprehensive server branding

### DISCUSSION THREAD

**Discussion by joshuaboniface** [jellyfin/jellyfin-meta#81] [Ideas]
Right now, it's fairly difficult to "brand" a server.

By brand, I mean replacing or adjusting the following things:

1. The logo that appears on the login and main screens.
2. The title of the instance (e.g. in the web `<title>` element)
3. The main colouration of the app.

Coupled to this, our "themes" right now are just changes to these things.

What I'd propose instead is simply allowing these to be configured directly in the server dashboard for each server instance.

The goals for this are twofold:

1. Allow server admins to customize their instances to their liking.
2. Allow us to somewhat enforce our branding guidelines on server admins who are doing things we don't like (e.g. paid services).

A benefit is then we could stop shipping "themes" that just change colours, and let these be tweaked ad-hoc by the admins.

I believe any sort of app branding (e.g. the icon you click on in Android for instance) is way beyond scope here; the idea would just be to brand the specific server and what is seen by a user after they connect to that server.

We currently have the branding API which returns the login screen disclaimer, so this would really just be an extension of that, permitting additional fields for the server name, logo, and basic colourization.

**Comment by thornbill** [comment]
1. The logo that appears on the login and main screens.
   * Would the favicon that shows in browser tabs for example remain unchanged? That would likely be challenging to customize.
2. The title of the instance (e.g. in the web `<title>` element)
   * Is this an additional field or just using the existing `ServerName` field from `SystemInfo`? In the dashboard and experimental layout we use this field currently.
4. The main colouration of the app.
   * I believe we would need:
     * Accent color -- would be used instead of our default blue
       * Since we use this color as the background for buttons we might also need a contrasting text color, white/black text option, or some way of determining which should be used. :thinking: 
     * Background color -- would be used instead of the default black-ish background
     * Light/Dark style setting -- could be used to determine font colors or tinting?

**Reply by sjorge** [reply]
> The title of the instance (e.g. in the web <title> element)

Perhaps just swap this to the server name? (Or something more fancy `<Show> - <Server name>` and `<Movie> - <Server name>` depending on if we're on a show or movie sub page.)

> Would the favicon that shows in browser tabs for example remain unchanged? That would likely be challenging to customize.

Would that not be about the same as changing the logo? The the tag in the `<head>` block would just need to change.

Unless you mean all the weird PWA icons too, that might indeed be a lot more difficult.

> The main colouration of the app. 

Swiftfin currently allows an accent color and it works pretty well, it does some fancy auto magic to pick white or black text for where text/icons appears inside those accent colors and the light or dark base also somehow influences that IIRC. Doing something similar might also work.


<details>
<summary>Just for reference, here are some screenshots from swiftfin with dark/light base and 2 random colors showing the text/icon color auto switching.</summary>

![IMG_2812](https://github.com/user-attachments/assets/e5971f7a-2ea6-4f09-b4ac-55486a401a8b)
![IMG_2811](https://github.com/user-attachments/assets/6ff3865f-6966-4934-8f8d-4d4cdf8f601c)
![IMG_2808](https://github.com/user-attachments/assets/5d49bdbe-03f1-40bc-8324-8a9ee0e3b9af)
![IMG_2807](https://github.com/user-attachments/assets/1872bbae-a3c5-49e0-b126-3c5749e25ef7)

</details>

**Reply by ferferga** [reply]
@thornbill 
> Would the favicon that shows in browser tabs for example remain unchanged?

The only requirement I see it's that server must be able to generate .ico files from images, which currently can't do afaik.

> Is this an additional field or just using the existing ServerName field from SystemInfo?

I'd say we should use the ``ServerName``, since that would cover most use cases already and minimize required changes.

> The main colouration of the app

I believe server should not dictate any color preference for dark/light theme, only the palette for each schema.
Ligh/dark theme should be handled client-side and stored by the client in `CustomPrefs` of the `DisplayPreferences` API, since it's the client the one that has abilities to check the OS' theme or user preferences, while the server data should be used as the baseline for the theme (I don't see this API overriding any further color customizations that the user might desire, in Vue I'd like to add color pickers for almost all the colors in the app: server palette would be the initial one and users will be able to customize based from those defaults, which right now are hardcoded into each app, the difference would be that the defaults are now hardcoded in server).

@sjorge The algorithm you're referring to and Swiftfin is probably using (but I don't know for sure) it's called APCA.

**Reply by lxqueen** [reply]
> The only requirement I see it's that server must be able to generate .ico files from images, which currently can't do afaik.

[PNG favicons are supported far enough back](https://caniuse.com/link-icon-png) that this shouldn't be a problem — could always include a vanilla Jellyfin `.ico` as the final fallback on the off chance an ancient browser or webOS _does_ somehow call for them.

**Reply by Hello-World-Traveler** [reply]
Using [imagemagick](https://www.imagemagick.org/) [convert](http://www.imagemagick.org/script/convert.php) 
`convert /logo.png -define icon:auto-resize=256,128,48,32,16 icon.ico`

Or using FFmpeg

`ffmpeg -i /logo.png -vf scale=32:32 /logo.ico
`

Not sure if this is helpful.
